### PR TITLE
Using attribute setters for maxpool

### PR DESCRIPTION
**Pull request by AlexandreEichenberger**
Use attribute setters for maxpool, this is the first of a series, just to iron out

### CODE REVIEW

File: src/dialect/onnx/onnx_ops.cpp [hunk 1]
```diff
@@ -1013,17 +1029,24 @@ void ONNXMaxPoolSingleOutOp::inferShapes() {
       actualPads.emplace_back(0);
     }
   }
+  // set pads
+  {
+    ArrayRef<int64_t> defaultRefs(actualPads);
+    auto defaultAttr = builder.getI64ArrayAttr(defaultRefs);
+    padsAttr(defaultAttr);
+    auto defaultAutoPadAttr = builder.getStringAttr("NOTSET");
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: I don't see you checking if this auto_pad already had a value. Looks like you're overwriting it.

Comment (doru1004) [reply src/dialect/onnx/onnx_ops.cpp]: Ah ok I see what's happening. Can you add a comment to explain why it looks like you're just overwriting it to this value?

File: src/dialect/onnx/onnx_ops.cpp [hunk 2]
```diff
@@ -885,11 +901,16 @@ void ONNXConvNoBiasOp::inferShapes() {
 //===----------------------------------------------------------------------===//
 
 // MaxPoolSingleOut
+// infer shape attributes output
+//    auto_pad set to NOTSET
+//    dilations, strides: set to 1 if not defined by user
+//    pads: set to proper value, 0 if not defined by user
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: Comment not formatted properly.

File: src/dialect/onnx/onnx_ops.cpp [hunk 3]
```diff
@@ -912,46 +932,44 @@ void ONNXMaxPoolSingleOutOp::inferShapes() {
...
+    // dilatation default is needed
+    SmallVector<int64_t, 4> defaultVals(kernelRank, 1);
+    // convert to ArrayRef, then build attribute, then store attribute
+    ArrayRef<int64_t> defaultRefs(defaultVals);
+    auto defaultAttr = builder.getI64ArrayAttr(defaultRefs);
+    dilationsAttr(defaultAttr);
+    dilationsOpt = dilations();
   }
 
   // storage order
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: Rogue comment?

File: src/dialect/onnx/onnx_ops.cpp [hunk 4]
```diff
@@ -912,46 +932,44 @@ void ONNXMaxPoolSingleOutOp::inferShapes() {
...
 
   // strides
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: Missing capital and full stop

File: src/dialect/onnx/onnx_ops.cpp [hunk 5]
```diff
@@ -912,46 +932,44 @@ void ONNXMaxPoolSingleOutOp::inferShapes() {
...
-  SmallVector<int64_t, 4> actualStrides;
   auto stridesOpt = strides();
   if (stridesOpt.hasValue()) {
-    auto stridesArray = stridesOpt.getValue().getValue();
-    if (stridesArray.size() != kernelRank)
+    if (ArrayAttrSize(stridesOpt) != kernelRank)
       emitError("strides rank is not the same as the spatial rank.");
     // fill in the actual values
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: Same.

File: src/dialect/onnx/onnx_ops.cpp [hunk 6]
```diff
@@ -912,46 +932,44 @@ void ONNXMaxPoolSingleOutOp::inferShapes() {
...
+      if (ArrayAttrIntVal(stridesOpt, i) < 1)
         emitError("strides value must be nonzero positive.");
-      actualStrides.emplace_back(s);
     }
   } else {
-    for (int i = 0; i < kernelRank; ++i) {
-      actualStrides.emplace_back(1);
-    }
+    SmallVector<int64_t, 4> defaultVals(kernelRank, 1);
+    // convert to ArrayRef, then build attribute, then store attribute
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: Same

File: src/dialect/onnx/onnx_ops.cpp [hunk 7]
```diff
@@ -912,46 +932,44 @@ void ONNXMaxPoolSingleOutOp::inferShapes() {
...
+    ArrayRef<int64_t> defaultRefs(defaultVals);
+    auto defaultAttr = builder.getI64ArrayAttr(defaultRefs);
+    stridesAttr(defaultAttr);
+    stridesOpt = strides();
   }
 
   // now try to find padding, getting auto_pad attribute first
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: Same.

File: src/dialect/onnx/onnx_ops.cpp [hunk 8]
```diff
@@ -962,13 +980,12 @@ void ONNXMaxPoolSingleOutOp::inferShapes() {
   if (autoPad == "NOTSET") {
     auto padsOpt = pads();
     if (padsOpt.hasValue()) {
-      auto padsArray = padsOpt.getValue().getValue();
       // pads consists of two entries for each spatial axis.
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: Same.

File: src/dialect/onnx/onnx_ops.cpp [hunk 9]
```diff
@@ -962,13 +980,12 @@ void ONNXMaxPoolSingleOutOp::inferShapes() {
...
-      if (padsArray.size() != 2 * kernelRank)
+      if (ArrayAttrSize(padsOpt) != 2 * kernelRank)
         emitError("pads rank is not twice the spatial rank.");
       // fill in the actual values
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: Same.

File: src/dialect/onnx/onnx_ops.cpp (hunk 1 above)
Comment (AlexandreEichenberger) [reply src/dialect/onnx/onnx_ops.cpp]: read on line 976. I want to overrite

File: src/dialect/onnx/onnx_ops.cpp (hunk 2 above)
Comment (AlexandreEichenberger) [reply src/dialect/onnx/onnx_ops.cpp]: it's a list... do you want me to add bullet points?

File: src/dialect/onnx/onnx_ops.cpp (hunk 3 above)
Comment (AlexandreEichenberger) [reply src/dialect/onnx/onnx_ops.cpp]: no, I don't handle it for now

Comment (AlexandreEichenberger) [reply src/dialect/onnx/onnx_ops.cpp]: added assert

File: src/dialect/onnx/onnx_ops.cpp (hunk 4 above)
Comment (AlexandreEichenberger) [reply src/dialect/onnx/onnx_ops.cpp]: done

**Pull request merged**